skdownscale/pointwise_models/quantile.py

```diff
@@ -0,0 +1,363 @@
+        if np.isinf(X_cdf.pp).any():
+            lower_inds = np.nonzero(-np.inf == X_cdf.pp)[0]
+            upper_inds = np.nonzero(np.inf == X_cdf.pp)[0]
+            model = LinearRegression()
+            if len(lower_inds):
+                s = slice(lower_inds[-1] + 1, lower_inds[-1] + 1 + self.n_endpoints)
+                model.fit(X_cdf.pp[s].reshape(-1, 1), X_cdf.vals[s].reshape(-1, 1))
+                X_cdf.pp[lower_inds] = model.predict(X_cdf.vals[lower_inds].reshape(-1, 1))
+            if len(upper_inds):
+                s = slice(upper_inds[0] - self.n_endpoints, upper_inds[0])
+                model.fit(X_cdf.pp[s].reshape(-1, 1), X_cdf.vals[s].reshape(-1, 1))
+                X_cdf.pp[upper_inds] = model.predict(X_cdf.vals[upper_inds].reshape(-1, 1))
+
+        # do the full quantile mapping
+        y_hat = np.full_like(X, np.nan)
+        y_hat[sort_inds] = np.interp(X_cdf.pp, self._y_cdf.pp, self._y_cdf.vals)[1:-1]
+
+        # If extrapolate is 1to1, apply the offset between ref and like to the
+        # tails of y_hat
+        if self.extrapolate == '1to1':
+            X_fit_len = len(self._X_cdf.vals)
+            X_fit_min = self._X_cdf.vals[0]
+            X_fit_max = self._X_cdf.vals[-1]
+
+            y_fit_len = len(self._y_cdf.vals)
+            y_fit_min = self._y_cdf.vals[0]
+            y_fit_max = self._y_cdf.vals[-1]
+
+            # adjust values over fit max
+            inds = X > X_fit_max
+            if inds.any():
+                if X_fit_len == y_fit_len:
+                    y_hat[inds] = y_fit_max + (X[inds] - X_fit_max)
+                elif X_fit_len > y_fit_len:
+                    X_fit_at_y_fit_max = np.interp(
+                        self._y_cdf.pp[-1], self._X_cdf.pp, self._X_cdf.vals
+                    )
+                    y_hat[inds] = y_fit_max + (X[inds] - X_fit_at_y_fit_max)
+                elif X_fit_len < y_fit_len:
+                    y_fit_at_X_fit_max = np.interp(
+                        self._X_cdf.pp[-1], self._y_cdf.pp, self._y_cdf.vals
+                    )
+                    y_hat[inds] = y_fit_at_X_fit_max + (X[inds] - X_fit_max)
+
+            # adjust values under fit min
+            inds = X < X_fit_min
+            if inds.any():
+                if X_fit_len == y_fit_len:
+                    y_hat[inds] = y_fit_min + (X[inds] - X_fit_min)
+                elif X_fit_len > y_fit_len:
+                    X_fit_at_y_fit_min = np.interp(
+                        self._y_cdf.pp[0], self._X_cdf.pp, self._X_cdf.vals
+                    )
+                    y_hat[inds] = X_fit_min + (X[inds] - X_fit_at_y_fit_min)
+                elif X_fit_len < y_fit_len:
+                    y_fit_at_X_fit_min = np.interp(
+                        self._X_cdf.pp[0], self._y_cdf.pp, self._y_cdf.vals
+                    )
+                    y_hat[inds] = y_fit_at_X_fit_min + (X[inds] - X_fit_min)
+
+        return y_hat
+
+    def _calc_extrapolated_cdf(
```

Comment:
brief comment describing what this does? I'll take a stab at it to save you time and maybe also can be a test of my own understanding:
'''
This function takes the tails of the cdf and extrapolates them out further via a linear model using the specified number of endpoints `n_endpoints`.
'''